[Doc][DocDB] improve architecture isolation levels page (#17666)
yugabyte/yugabyte-db: docs/content/preview/architecture/transactions/isolation-levels.md

## Locking granularities

Locks can be taken at many levels of granularity.  For example, a serializable read lock could be taken at the level of an entire tablet, a single row, or a single column of a single row.  Such a lock will block attempts to take write locks at that or lower granularities. Thus, for example, a read lock taken at the row level will block attempts to write to that entire row or any sub column of it.

In addition to the above-mentioned levels of granularity, locks in DocDB can be taken at prefixes of the primary key, treating the hash columns as a single unit.  For example, if we have created a YSQL table via:
```sql
CREATE TABLE test (h1 INT, h2 INT, r1 INT, r2 INT, v INT w INT PRIMARY KEY ((h1,h2) HASH, r1 ASC, r2 ASC);
```

then we could lock at any of the following:
- the entire tablet
- all rows having h1=2, h2=3
- all rows having h1=2, h2=3, r1=4
- the row having h1=2, h2=3, r1=4, r2=5
- column v of the row having h1=2, h2=3, r1=4, r2=5

> With YCQL, granularities exist below the column level; for example only one key of a column map can be locked.

Reviewer: ```suggestion
With YCQL, granularities exist below the column level; for example, only one key of a column map can be locked.
```

Author: :+1:

## Efficiently detecting conflicts between locks of different granularities

The straightforward way to handle locks of different granularities would be to have a map from location to lock types.  However, this is too inefficient for detecting conflicts: attempting, for example, to add a lock at the tablet level would require checking for locks at every row and column in that tablet.

> To make conflict detection efficient, YugabyteDB stores extra information at each location about the locks at lower granularities below it.  In particular, it takes a normal lock at the original granularity and weaker versions of that lock at all the granularities that enclose the original granularity.  We call the normal locks _strong_ locks and the weaker variants _weak_ locks.

Reviewer: ```suggestion
To make conflict detection efficient, YugabyteDB stores extra information at each location about the locks at granularities below it.  In particular, it takes a normal lock at the original granularity and weaker versions of that lock at all the granularities that enclose the original granularity. The normal locks are called _strong_ locks and the weaker variants _weak_ locks.
```

Author: style guide says to prefer active voice

Author: good: removing "lower", however

Reviewer: Style guide not mention we don't use we, I forget. But we don't.

Author: :+1:

> As an example, pretend we have only tablet- and row-level granularities. To take a serializable write lock at the row level, we would take a strong write lock at the row level and a weak write lock at the tablet level.  To take a serializable read lock at the tablet level, we would just take a strong read lock at the tablet level.

Reviewer: ```suggestion
As an example, suppose you have only tablet- and row-level granularities. To take a serializable write lock at the row level, you would take a strong write lock at the row level and a weak write lock at the tablet level.  To take a serializable read lock at the tablet level, you would just take a strong read lock at the tablet level.
```

Author: The current version reads better to me because the connotations are that  we are temporarily changing the system.  Somehow I don't feel "you" has permission to do that.

Reviewer: What's really doing the work here - the database? As noted, 'we' is typically avoided, but I'm not that fussed about it.

> By using the following conflict rules, we can decide if two original locks would conflict based only on whether or not their strong/weak locks at any location would conflict:

Reviewer: ```suggestion
Using the following conflict rules, you can decide if two original locks would conflict based only on whether or not their strong/weak locks at any location would conflict:
```

Author: dumping the "by",  but I think we is better here because it includes the system unlike "you".

> - two strong locks conflict iff they would've conflicted as normal locks

Reviewer: ```suggestion
- two strong locks conflict if they would've conflicted as normal locks
```

Author: that was not a typo,  but I will spell it out to avoid confusion (if and only if)

Reviewer: another one for the style guide - spell out acronyms I've never heard of heh

Author: technically an abbreviation and an extremely common one in math at that but who knows what our readers are familiar with

Reviewer: is no abbr. not in the style guide? no etc, eg, ie, etc. spell it all out thx!

- two weak locks never conflict

> - a strong lock conflicts with a weak lock iff they would've conflicted as normal locks

Reviewer: ```suggestion
- a strong lock conflicts with a weak lock if they would've conflicted as normal locks
```

Author: ditto

Reviewer: ditto

That is, for each location that would have two locks, would they conflict under the above rules?  There is no need to enumerate the sublocations of any location.

> In our example, we would detect a conflict at the tablet level between the strong read lock and the weak write lock.  If we had instead tried to take the serializable read lock at the row level then there would've been no conflict at the tablet level (both locks there are weak) but there might have been at the row level if the two locks were taken on the same row.

Reviewer: ```suggestion
In the example, you would detect a conflict at the tablet level between the strong read lock and the weak write lock.  If you had instead tried to take the serializable read lock at the row level, then there would've been no conflict at the tablet level (both locks there are weak) but there might have been at the row level if the two locks were taken on the same row.
```

Author: made changes except left it as we because of the including the system

> When we include the strong/weak distinction, the full conflict matrix becomes:

Reviewer: ```suggestion
Including the strong/weak distinction, the full conflict matrix becomes:
```

Author: :+1:
